### PR TITLE
Deallocate addresses to return to pool

### PR DESCRIPTION
**Pull request by debedb**
https://paninetworks.kanbanize.com/ctrl_board/3/cards/286/details

### CODE REVIEW

File: ipam/ipam.go [hunk 1]
```diff
@@ -40,12 +40,19 @@ func (ipam *IPAMSvc) Routes() common.Routes {
 	routes := common.Routes{
 		common.Route{
 			Method:          "POST",
-			Pattern:         "/vms",
-			Handler:         ipam.addVm,
-			MakeMessage:     func() interface{} { return &Vm{} },
+			Pattern:         "/endpoints",
```
Comment (debedb) [inline ipam/ipam.go]: The OpenStack plugin still uses the legacy one. Checking about the K8S one.

File: common/client.go [hunk 2]
```diff
@@ -245,7 +245,7 @@ func (rc *RestClient) execMethod(method string, dest string, data interface{}, r
 			return err
 		}
 		req.Header.Set("accept", "application/json")
-
+
```
Comment (jbrendel) [inline common/client.go]: Leading spaces in an empty line?

File: common/client.go [hunk 3]
```diff
@@ -318,6 +318,12 @@ func (rc *RestClient) Post(url string, data interface{}, result interface{}) err
 	return err
 }
 
+// Post applies DELETE method to the specified URL
```
Comment (jbrendel) [inline common/client.go]: I think the comment should start with 'Delete'

File: ipam/ipam.go [hunk 4]
```diff
@@ -146,26 +153,21 @@ func (ipam *IPAMSvc) legacyAllocateIpByName(input interface{}, ctx common.RestCo
...
 		return nil, errors.New("Segment with name " + hostName + " not found")
 	}
-	log.Printf("Sement name %s has ID %s", segmentName, vm.SegmentId)
+	log.Printf("Segment name %s has ID %s", segmentName, Endpoint.SegmentId)
 
-	return ipam.addVm(&vm, ctx)
+	return ipam.addEndpoint(&Endpoint, ctx)
 }
 
 // handleHost handles request for a specific host's info
```
Comment (jbrendel) [inline ipam/ipam.go]: Comment should start with 'addEndpoint'

File: ipam/store.go [hunk 5]
```diff
@@ -17,85 +17,130 @@ package ipam
...
+	"errors"
+	"fmt"
 	"github.com/romana/core/common"
 	"log"
+	"strings"
 )
 
-type Vm struct {
-	//	Id        uint64 `json:"id"`
+type Endpoint struct {
```
Comment (jbrendel) [inline ipam/store.go]: This structure deserves a docstring

File: ipam/store.go [hunk 6]
```diff
@@ -17,85 +17,130 @@ package ipam
...
+	Endpoints []Endpoint
+	Id        string `sql:"unique_index"`
 }
 
 type ipamStore struct {
 	common.DbStore
 }
 
-func (ipamStore *ipamStore) addVm(stride uint, vm *Vm) error {
+func (ipamStore *ipamStore) deleteEndpoint(ip string) (Endpoint, error) {
```
Comment (jbrendel) [inline ipam/store.go]: docstring

File: common/client.go (hunk 2 above)
Comment (pritesh) [inline common/client.go]: "make fmt && make vet" is your friend.

File: ipam/store.go [hunk 7]
```diff
@@ -17,85 +17,130 @@ package ipam
...
 	err := common.MakeMultiError(tx.GetErrors())
 	if err != nil {
 		tx.Rollback()
+		return Endpoint{}, err
+	}
+	tx.Commit()
+	return results[0], nil
+}
+
+func (ipamStore *ipamStore) addEndpoint(endpoint *Endpoint, upToEndpointIpInt uint64, stride uint) error {
```
Comment (jbrendel) [inline ipam/store.go]: docstring

File: ipam/store.go [hunk 8]
```diff
@@ -17,85 +17,130 @@ package ipam
...
+	segId := endpoint.SegmentId
+
+	where := "host_id = ? AND segment_id = ? AND in_use = 0"
+	sel := "min(seq)"
+	log.Printf("Calling SELECT %s FROM endpoints WHERE %s;", sel, fmt.Sprintf(strings.Replace(where, "?", "%s", 2), hostId, segId))
+	row := tx.Model(Endpoint{}).Where(where, hostId, segId).Select(sel).Row()
+	seq := sql.NullInt64{}
+	row.Scan(&seq)
+	log.Printf("minseq: %v", seq)
+	if !seq.Valid {
```
Comment (jbrendel) [inline ipam/store.go]: I would love to see some comments scattered throughout this function to explain what's going on. There is some involved logic and arithmetic, which would be well served with a few explanations.

File: ipam/store.go [hunk 9]
```diff
@@ -17,85 +17,130 @@ package ipam
...
-	log.Printf("New sequence is %d\n", vm.Seq)
-
-	// vmSeq is the sequence number of VM in a given host
-	effectiveVmSeq := getEffectiveSeq(vm.Seq, stride)
-	log.Printf("Effective sequence for seq %d (stride %d): %d\n", vm.Seq, stride, effectiveVmSeq)
-	vm.EffectiveSeq = effectiveVmSeq
-	ipamVm := IpamVm{Vm: *vm}
-	tx.NewRecord(ipamVm)
-	tx = tx.Create(&ipamVm)
+	tx = tx.Model(Endpoint{}).Where("ip = ?", ip).Update("in_use", false)
```
Comment (jbrendel) [inline ipam/store.go]: So, this function doesn't actually delete anything, it just marks things as not in use anymore? Might be worth a comment.

Are those things ever actually deleted?

File: ipam/store.go [hunk 10]
```diff
@@ -17,85 +17,130 @@ package ipam
...
+	if err != nil {
+		tx.Rollback()
 		return err
 	}
 	tx.Commit()
 	return nil
 }
 
-// getEffectiveSeq gets effective sequence number of a VM
+// getEffectiveSeq gets effective sequence number of a Endpoint
```
Comment (jbrendel) [inline ipam/store.go]: "an Endpoint"

Also, might be worth explaining why you call it the 'effective' sequence number. Is there an in-effective one? Or a different one that's not ... effective?

**Comment by jbrendel** [comment]
So, who initiates the deletion of endpoints?

OpenStack when a VM is brought down/deleted?

Kubernetes when a pod is destroyed?

What happens if a host is deleted? Do the environments (OS or k8s) send individual delete events for each VM?

What if the host crashes? Do the environments eventually detect this and then start generating delete events?

File: common/client.go (hunk 2 above)
Comment (debedb) [inline common/client.go]: Yes, will do!

File: common/client.go (hunk 3 above)
Comment (debedb) [inline common/client.go]: duh

File: ipam/ipam.go (hunk 4 above)
Comment (debedb) [inline ipam/ipam.go]: yes

File: ipam/store.go (hunk 5 above)
Comment (debedb) [inline ipam/store.go]: ok

File: ipam/store.go (hunk 6 above)
Comment (debedb) [inline ipam/store.go]: ok

File: ipam/store.go (hunk 7 above)
Comment (debedb) [inline ipam/store.go]: ok

File: ipam/store.go (hunk 9 above)
Comment (debedb) [inline ipam/store.go]: Should be deleted only when a parent entity (segment or host) is deleted.

File: ipam/ipam.go [hunk 11]
```diff
@@ -146,26 +153,21 @@ func (ipam *IPAMSvc) legacyAllocateIpByName(input interface{}, ctx common.RestCo
...
-func (ipam *IPAMSvc) addVm(input interface{}, ctx common.RestContext) (interface{}, error) {
-	vm := input.(*Vm)
-	err := ipam.store.addVm(ipam.dc.EndpointSpaceBits, vm)
-	if err != nil {
-		log.Printf("HEYHEYHEY %v\n", err)
-		return nil, err
-	}
+func (ipam *IPAMSvc) addEndpoint(input interface{}, ctx common.RestContext) (interface{}, error) {
```
Comment (jbrendel) [inline ipam/ipam.go]: We should have a few comments throughout this function. It's long and complex enough.

File: ipam/store.go (hunk 9 above)
Comment (jbrendel) [inline ipam/store.go]: Ok...? Where does this happen? And maybe just explain this here in a comment.

Comment (debedb) [inline ipam/store.go]: That's better dealt with after https://paninetworks.kanbanize.com/ctrl_board/3/cards/296/details

File: ipam/store.go (hunk 10 above)
Comment (debedb) [inline ipam/store.go]: It's explained in the struct. Added.

**Comment by debedb** [comment]
Test

**Comment by debedb** [comment]
Not sure how to answer it up above so will do here:

> So, who initiates the deletion of endpoints?
> OpenStack when a VM is brought down/deleted?

Yes. 

> Kubernetes when a pod is destroyed?

I would guess so - @flashvoid , what happens now?

> What happens if a host is deleted? Do the environments (OS or k8s) send individual delete 
> events for each VM?

I was thinking of having another method of deleting host or tenant or segment from IPAM, which would delete all information. But that's better done after https://paninetworks.kanbanize.com/ctrl_board/3/cards/296/details

> What if the host crashes? Do the environments eventually detect this and then start generating 
> delete events?

Does OpenStack even do such a thing?

**Pull request merged**